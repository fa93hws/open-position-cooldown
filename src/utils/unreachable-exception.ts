export class UnreachableException extends Error {
  constructor(val: never) {
    super(`unreachable value: ${JSON.stringify(val)}`);
  }
}
