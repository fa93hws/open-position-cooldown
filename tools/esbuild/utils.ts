import * as fs from 'fs';
import * as path from 'path';
import { OutputFile } from 'esbuild';
import * as crypto from 'crypto';
import { template } from 'lodash';

export const enum BUILD_MODE {
  PROD = 'production',
  DEV = 'development',
}

export function ensureFolder(target: string): void {
  fs.rmdirSync(target, { recursive: true });
  fs.mkdirSync(target, { recursive: true });
}

export function hashOutputs(outputFiles: readonly OutputFile[]): OutputFile[] {
  return outputFiles.map((file) => {
    let filePath: string;
    if (['.js', '.css'].includes(path.extname(file.path))) {
      const md5sum = crypto.createHash('md5');
      const filename = md5sum.update(file.contents).digest('hex').slice(0, 16);
      filePath = path.join(
        path.dirname(file.path),
        filename + path.extname(file.path),
      );
    } else {
      filePath = file.path;
    }
    return {
      path: filePath,
      contents: file.contents,
    };
  });
}

function assertExists({
  filePath,
  name,
  existsSync = fs.existsSync,
}: {
  filePath: string;
  name: string;
  existsSync?: typeof fs.existsSync;
}) {
  if (!existsSync(filePath)) {
    throw new Error(`${name} ${filePath} does not exist`);
  }
}

export function generateHtml({
  templatePath,
  files,
  outdir,
  existsSync = fs.existsSync,
  writeFileSync = fs.writeFileSync,
}: {
  templatePath: string;
  files: readonly string[];
  outdir: string;
  existsSync?: typeof fs.existsSync;
  writeFileSync?: typeof fs.writeFileSync;
}) {
  assertExists({ filePath: templatePath, name: 'templatePath', existsSync });
  assertExists({ filePath: outdir, name: 'outdir', existsSync });
  files.forEach((file, idx) =>
    assertExists({ filePath: file, name: `file[${idx}]`, existsSync }),
  );
  const templateContent = fs.readFileSync(templatePath, { encoding: 'utf-8' });
  const jsFiles = files
    .filter((f) => path.extname(f) === '.js')
    .map((f) => path.relative(outdir, f));
  const cssFiles = files
    .filter((f) => path.extname(f) === '.css')
    .map((f) => path.relative(outdir, f));
  const htmlOutputContent = template(templateContent)({ jsFiles, cssFiles });
  writeFileSync(path.join(outdir, 'index.html'), htmlOutputContent);
}

export function copyAssets({
  from,
  to,
  copyFileSync = fs.copyFileSync,
}: {
  from: string;
  to: string;
  copyFileSync?: typeof fs.copyFileSync;
}) {
  assertExists({ filePath: from, name: 'from' });
  ensureFolder(to);
  fs.readdirSync(from, { withFileTypes: true })
    .filter((dirent) => dirent.isFile())
    .map((dirent) => dirent.name)
    .forEach((file) =>
      copyFileSync(path.join(from, file), path.join(to, file)),
    );
}
