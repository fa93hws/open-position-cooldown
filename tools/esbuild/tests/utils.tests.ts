import * as path from 'path';
import { TextEncoder } from 'util';

import { hashOutputs, generateHtml } from '../utils';

describe('hashOutputs', () => {
  const enc = new TextEncoder();

  it('hash the filename with its content by md5', () => {
    const contents = enc.encode('123');
    const outputs = [
      {
        path: 'a/b/c/d.js',
        contents,
      },
      {
        path: 'b/e/d.sss.css',
        contents,
      },
    ];
    const hashedOutput = hashOutputs(outputs);
    expect(hashedOutput).toEqual([
      {
        path: 'a/b/c/202cb962ac59075b.js',
        contents,
      },
      {
        path: 'b/e/202cb962ac59075b.css',
        contents,
      },
    ]);
  });

  it('will not hash the filename for font file', () => {
    const contents = enc.encode('123');
    const outputs = [
      {
        path: 'a/b/c/d.woff',
        contents,
      },
      {
        path: 'b/e/d.sss.woff2',
        contents,
      },
    ];
    const hashedOutput = hashOutputs(outputs);
    expect(hashedOutput).toEqual(outputs);
  });
});

describe('generateHtml', () => {
  it('injects files into template', () => {
    const outdir = path.join(__dirname, 'fixtures');
    const templatePath = path.join(outdir, 'baz.ejs');
    const files = [
      path.join(outdir, 'foo', 'a.js'),
      path.join(outdir, 'b.css'),
      path.join(outdir, 'c.js'),
      path.join(outdir, 'bar', 'd.css'),
    ];

    const existsSync = () => true;
    const writeFileSync = jest.fn();

    generateHtml({
      templatePath,
      files,
      outdir,
      existsSync,
      writeFileSync,
    });

    expect(writeFileSync.mock.calls[0][1]).toMatchSnapshot();
    expect(writeFileSync.mock.calls[0][0]).toEqual(
      path.join(outdir, 'index.html'),
    );
  });
});
