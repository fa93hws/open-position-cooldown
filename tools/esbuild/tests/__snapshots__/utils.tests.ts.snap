// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateHtml injects files into template 1`] = `
<link rel="stylesheet"
      href="b.css"
>
<link rel="stylesheet"
      href="bar/d.css"
>
<script src="foo/a.js">
</script>
<script src="c.js">
</script>
`;
